### PR TITLE
Removed discount strategy and holding pen

### PR DESCRIPTION
**Pull request by aalavandhan**
* Removed holding pen from perp
* Removed discount strategy and associated math. Perp only accepts the senior most tranches (instead of checking for discounts).
* Updated fee strategy calculation, removed effective discount adjusted ratio logic. Target [spot, stAMPL ratio] is expected to be the tranche ratio of the senior most tranche of the minting bond.

### CODE REVIEW

File: spot-contracts/contracts/PerpetualTranche.sol [hunk 1]
```diff
@@ -911,63 +823,47 @@ contract PerpetualTranche is
...
-        // Given the amount of tranches In, we compute the amount of tranches out
+        // Given the amount of tranches In, we compute the amount of tokens out
         r.trancheInAmt = trancheInAmtAvailable;
-        uint256 stdTrancheInAmt = _toStdTrancheAmt(trancheInAmtAvailable, trancheInDiscount);
-        uint256 stdTrancheOutAmt = stdTrancheInAmt.mulDiv(trancheInPrice, trancheOutPrice);
+        r.tokenOutAmt = r.trancheInAmt.mulDiv(trancheInPrice, tokenOutPrice);
 
-        // A positive fee percentage implies that perp charges rotators by
-        // accepting tranchesIn at a discount, ie) fewer tranches out.
+        // A postive fee percentage implies that perp charges rotators by
```
Comment (brandoniles) [inline spot-contracts/contracts/PerpetualTranche.sol:844]:
```suggestion
        // A positive fee percentage implies that perp charges rotators by
```

File: spot-contracts/contracts/PerpetualTranche.sol [hunk 2]
```diff
@@ -911,63 +823,47 @@ contract PerpetualTranche is
...
-        bool isMatureTrancheOut = _isMatureTranche(tokenOut);
-        r.tokenOutAmt = isMatureTrancheOut
-            ? tokenOutBalance.mulDiv(r.trancheOutAmt, _matureTrancheBalance)
-            : r.trancheOutAmt;
-
         // When the tokenOut balance is NOT covered:
         // we fix tokenOutAmt = tokenOutAmtRequested and re-calculate other values
-        if (r.tokenOutAmt > tokenOutAmtRequested) {
-            // Given the amount of tranches out, we compute the amount of tranches in
+        if (r.tokenOutAmt >= tokenOutAmtRequested) {
```
Comment (brandoniles) [inline spot-contracts/contracts/PerpetualTranche.sol:858]: Is the `=` in`>=` correct? Sounds like we need to continue only if there'll be tokens left over

File: spot-contracts/contracts/PerpetualTranche.sol [hunk 3]
```diff
@@ -593,30 +532,20 @@ contract PerpetualTranche is
             type(uint256).max
         );
 
-        // verifies if rollover amount is acceptable
-        if (r.trancheInAmt == 0 || r.tokenOutAmt == 0 || r.perpRolloverAmt == 0) {
+        // Verifies if rollover amount is acceptable
+        if (r.trancheInAmt == 0) {
```
Comment (brandoniles) [inline spot-contracts/contracts/PerpetualTranche.sol:536]: Is there motivation for removing the `tokenOutAmt` check?

File: spot-contracts/contracts/PerpetualTranche.sol [hunk 4]
```diff
@@ -752,6 +674,19 @@ contract PerpetualTranche is
         return _rollers.at(i);
     }
 
+    /// @inheritdoc IPerpetualTranche
+    function computeDiscount(IERC20Upgradeable token) external view override returns (uint256) {
```
[the remark this answers is not in the document]
Comment (aalavandhan) [reply spot-contracts/contracts/PerpetualTranche.sol:678]: We can do a final pass to remove all unused interface functions.

File: spot-contracts/contracts/PerpetualTranche.sol (hunk 3 above)
Comment (aalavandhan) [reply spot-contracts/contracts/PerpetualTranche.sol:536]: From the spot contract perspective, we care more about tokens coming it. 

ie) trancheInAmt > 0 and tokenOutAmt >= 0 is reasonable. However trancheInAmt = 0 is unreasonable.

File: spot-contracts/contracts/strategies/TrancheClassDiscountStrategy.sol [hunk 5]
```diff
@@ -1,80 +0,0 @@
...
- *        - it's seniorityIDX.
- *
- *       For example:
- *        - All AMPL [35-65] bonds can be configured to have a discount of [1, 0] and
- *        => An AMPL-A tranche token from any [35-65] bond will be applied a discount factor of 1.
- *        - All AMPL [50-50] bonds can be configured to have a discount of [0.8,0]
- *        => An AMPL-A tranche token from any [50-50] bond will be applied a discount factor of 0.8.
- *
- */
-contract TrancheClassDiscountStrategy is IDiscountStrategy, OwnableUpgradeable {
```
[the remark this answers is not in the document]
Comment (aalavandhan) [reply spot-contracts/contracts/strategies/TrancheClassDiscountStrategy.sol:29]: It still being used by the interface method `computeDiscount`.. Will do final pass to remove unused code

File: spot-contracts/contracts/PerpetualTranche.sol (hunk 2 above)
Comment (aalavandhan) [reply spot-contracts/contracts/PerpetualTranche.sol:858]: Right we don't need to re-compute when `r.tokenOutAmt == tokenOutAmtRequested`

File: spot-contracts/contracts/_interfaces/IPerpetualTranche.sol [hunk 6]
```diff
@@ -81,12 +63,12 @@ interface IPerpetualTranche is IERC20Upgradeable {
...
-    /// @return Address of the collateral token.
+    /// @return Address of the underlying collateral token.
     function collateral() external view returns (IERC20Upgradeable);
 
-    /// @notice The "virtual" balance of all mature tranches held by the system.
-    /// @return The mature tranche balance.
-    function getMatureTrancheBalance() external returns (uint256);
+    /// @notice The address of the underlying rebasing ERC-20 collateral token backing the tranches.
+    /// @return Address of the underlying collateral token.
+    function underlying() external view returns (IERC20Upgradeable);
```
[the remark this answers is not in the document]
Comment (aalavandhan) [reply spot-contracts/contracts/_interfaces/IPerpetualTranche.sol:71]: I favor `underlying` over `collateral`. Lets keep it for now but remove in a later pass

File: spot-contracts/contracts/PerpetualTranche.sol (hunk 3 above)
Comment (aalavandhan) [reply spot-contracts/contracts/PerpetualTranche.sol:536]: On second thought, adding it back for the sake of consistency. In other functions like deposit we check for either the note == 0 or the underlyingIn == 0 to revert.

Comment (aalavandhan) [reply spot-contracts/contracts/PerpetualTranche.sol]: Makes sense

File: spot-contracts/contracts/RouterV2.sol [hunk 7]
```diff
@@ -59,77 +59,37 @@ contract RouterV1 {
...
-            if (mintAmt <= 0) {
-                // transfers unused tranches back
-                bt.tranches[i].safeTransfer(msg.sender, trancheAmt);
-            }
+        // transfers remaining tranches back
+        for (uint8 i = 1; i < bt.tranches.length; i++) {
+            bt.tranches[i].safeTransfer(msg.sender, bt.tranches[i].balanceOf(address(this)));
         }
 
         // transfers any remaining collateral tokens back
```
Comment (nms-7) [inline spot-contracts/contracts/RouterV2.sol:95]: I think we can safely remove this transfer back

Comment (nms-7) [inline spot-contracts/contracts/RouterV2.sol:95]: I think we can safely remove transferring back remaining collateral here

Comment (aalavandhan) [reply spot-contracts/contracts/RouterV2.sol:95]: Will make a note, will remove as part of the gas efficiency pass later

Comment (aalavandhan) [inline spot-contracts/contracts/RouterV2.sol:95]:
```suggestion
        // TODO: revisit removing this 
        // transfers any remaining collateral tokens back
```

**Review by brandoniles — APPROVED** [review]
LGTM

**Review by nms-7 — APPROVED** [review]

**Pull request merged**